### PR TITLE
[Proposal] Lazy compactMap sequence

### PR DESCRIPTION
**Pull request by tellowkrinkle**
This is the formal proposal for [this thread](https://forums.swift.org/t/introduce-lazy-version-of-compactmap/9835)

### CODE REVIEW

File: proposals/NNNN-lazy-compactmap-sequence.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+In addition, code like following code will break:
+```swift
+let array = [0, 1, 22]
+let tmp = array.lazy.map(String.init).filter { $0.count == 1 }
+let filtered: LazyFilterCollection<LazyMapCollection<[Int], String>> = tmp
+```
+
+However, this type of code is probably rare and similar code will already
+be broken by the previously mentioned change that coalesces
+`.filter(_:).filter(_:)` and `.map(_:).map(_:)`
```
[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply proposals/NNNN-lazy-compactmap-sequence.md]: This code compiles on 4.0.3 but not on master, so I would say that means that it's not source compatible (and is an example of what I was describing in the document):
```swift
func test() -> LazyFilterBidirectionalCollection<LazyFilterBidirectionalCollection<[Int]>> {
	let a = [2]
	let b = a.lazy.filter({ $0 == 2 }).filter({ $0 == 2 })
	return b
}
```

[the remark this answers is not in the document]
Comment (moiseev) [reply proposals/NNNN-lazy-compactmap-sequence.md]: It just means there is a compatibility bug... Thanks for reporting it!

[the remark this answers is not in the document]
Comment (moiseev) [reply proposals/NNNN-lazy-compactmap-sequence.md]: I believe [this](https://github.com/apple/swift/pull/14859) should do it. Thanks again, @tellowkrinkle

[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply proposals/NNNN-lazy-compactmap-sequence.md]: Ahh, I see.  I'll make similar changes to my implementation.

In that case, should the `Source Compatibility` section cover source compatibility in Swift 4 mode, in the transition to Swift 5, or both?

[the remark this answers is not in the document]
Comment (moiseev) [reply proposals/NNNN-lazy-compactmap-sequence.md]: Let me figure this out for you...

[the remark this answers is not in the document]
Comment (tkremenek) [reply proposals/NNNN-lazy-compactmap-sequence.md]: @tellowkrinkle it should cover both.  If there is a source-breaking change here, it needs to be clearly outlined, motivated, and described the intended rollout (e.g., keep compatibility in Swift 4 mode, roll out in Swift 5).  Source-breaking changes need to meet a very high bar.

[the remark this answers is not in the document]
Comment (tellowkrinkle) [reply proposals/NNNN-lazy-compactmap-sequence.md]: Alright I've updated the implementation and proposal to be fully compatible in Swift 4 mode

**Comment by rjmccall** [comment]
Sorry for the long delay.  We're thinking of scheduling this for July 30th – August 5th; does that work for you?

**Comment by tellowkrinkle** [comment]
Yeah that should be fine

**Pull request merged**